crates/bevy_pbr/src/render_graph/forward_pipeline/forward.frag

```diff
@@ -24,41 +27,170 @@ layout(set = 1, binding = 0) uniform Lights {
+    return attenuation * 1.0 / max(distanceSquare, 1e-4);
+}
+
+float D_GGX(float roughness, float NoH, const vec3 h) {
+    float oneMinusNoHSquared = 1.0 - NoH * NoH;
+    float a = NoH * roughness;
+    float k = roughness / (oneMinusNoHSquared + a * a);
+    float d = k * k * (1.0 / PI);
+    return d;
+}
+
+float V_SmithGGXCorrelated(float roughness, float NoV, float NoL) {
+    float a2 = roughness * roughness;
+    float lambdaV = NoL * sqrt((NoV - a2 * NoV) * NoV + a2);
+    float lambdaL = NoV * sqrt((NoL - a2 * NoL) * NoL + a2);
+    float v = 0.5 / (lambdaV + lambdaL);
+    return v;
+}
+
+vec3 F_Schlick(const vec3 f0, float f90, float VoH) {
+    return f0 + (f90 - f0) * pow5(1.0 - VoH);
+}
+
+float F_Schlick(float f0, float f90, float VoH) {
+    return f0 + (f90 - f0) * pow5(1.0 - VoH);
+}
+
+vec3 fresnel(vec3 f0, float LoH) {
+    float f90 = saturate(dot(f0, vec3(50.0 * 0.33)));
+    return F_Schlick(f0, f90, LoH);
+}
+
+vec3 isotropicLobe(vec3 f0, float roughness, const vec3 h, float NoV, float NoL,
+                   float NoH, float LoH) {
+
+    float D = D_GGX(roughness, NoH, h);
+    float V = V_SmithGGXCorrelated(roughness, NoV, NoL);
+    vec3 F = fresnel(f0, LoH);
+
+    return (D * V) * F;
+}
+
+float Fd_Burley(float roughness, float NoV, float NoL, float LoH) {
+    float f90 = 0.5 + 2.0 * roughness * LoH * LoH;
+    float lightScatter = F_Schlick(1.0, f90, NoL);
+    float viewScatter = F_Schlick(1.0, f90, NoV);
+    return lightScatter * viewScatter * (1.0 / PI);
+}
+
+vec3 computeDiffuseColor(const vec3 baseColor, float metallic) {
+    return baseColor * (1.0 - metallic);
+}
+
+#define MIN_N_DOT_V 1e-4
+
+float clampNoV(float NoV) {
+    // Neubelt and Pettineo 2013, "Crafting a Next-gen Material Pipeline for The
+    // Order: 1886"
+    return max(NoV, MIN_N_DOT_V);
+}
+
+float perceptualRoughnessToRoughness(float perceptualRoughness) {
+    return perceptualRoughness * perceptualRoughness;
+}
+
+#if defined(TARGET_MOBILE)
+// min roughness such that (MIN_PERCEPTUAL_ROUGHNESS^4) > 0 in fp16 (i.e.
+// 2^(-14/4), rounded up)
+#define MIN_PERCEPTUAL_ROUGHNESS 0.089
+#define MIN_ROUGHNESS 0.007921
+#else
+#define MIN_PERCEPTUAL_ROUGHNESS 0.045
+#define MIN_ROUGHNESS 0.002025
+#endif
+
+#endif
 
 void main() {
     vec4 output_color = Albedo;
-# ifdef STANDARDMATERIAL_ALBEDO_TEXTURE
-    output_color *= texture(
-        sampler2D(StandardMaterial_albedo_texture, StandardMaterial_albedo_texture_sampler),
-        v_Uv);
-# endif
-
-# ifndef STANDARDMATERIAL_UNLIT
-    vec3 normal = normalize(v_Normal);
+#ifdef STANDARDMATERIAL_ALBEDO_TEXTURE
+    output_color *= texture(sampler2D(StandardMaterial_albedo_texture,
+                                      StandardMaterial_albedo_texture_sampler),
+                            v_Uv);
+#endif
+
+#ifndef STANDARDMATERIAL_UNLIT
+    float perceptual_roughness = pbr.x;
+    perceptual_roughness =
+        clamp(perceptual_roughness, MIN_PERCEPTUAL_ROUGHNESS, 1.0);
```

Comment:
Long term we probably should have some kind of API for doing this sort of uniform value range restriction outside the shader. We could also do the same for packing a bunch of single value floats into a vector.

Reply:
Could do this, or even with an enum that allows linear and non-linear roughness to be used? This way the shader doesn't need to compute non-linear roughness anymore and we can even warn the user when their input is clamped.

In `crates\bevy_pbr\src\material.rs`:
```rust
const MIN_ROUGHNESS: f32 = 0.089;
const MAX_ROUGHNESS: f32 = 1.0;

#[derive(Debug)]
pub struct PerceptualRoughness(pub f32);

impl RenderResource for PerceptualRoughness {
    fn resource_type(&self) -> Option<bevy_render::renderer::RenderResourceType> {
        Some(RenderResourceType::Buffer)
    }

    fn write_buffer_bytes(&self, buffer: &mut [u8]) {
        // Clamp and precompute non-linear roughness
        // TODO or constructor that doesn't take out of range values
        let mut roughness = self.0.max(MIN_ROUGHNESS).min(MAX_ROUGHNESS);
        roughness *= roughness;
        roughness.write_bytes(buffer);
    }

    fn buffer_byte_len(&self) -> Option<usize> {
        Some(self.0.byte_len())
    }

    fn texture(&self) -> Option<&Handle<Texture>> {
        None
    }
}

/// A material with "standard" properties used in PBR lighting
/// Standard property values with pictures here https://google.github.io/filament/Material%20Properties.pdf
#[derive(Debug, RenderResources, ShaderDefs, TypeUuid)]
#[uuid = "dace545e-4bc6-4595-a79d-c224fc694975"]
pub struct StandardMaterial {
    /// Doubles as diffuse albedo for non-metallic, specular for metallic and a mix for everything in between
    pub base_color_factor: Color,
    #[shader_def]
    pub base_color_texture: Option<Handle<Texture>>,
    /// Linear perceptual roughness, clamped to [0.089,1] in the shader
    /// Defaults to minimum of 0.089
    pub roughness_factor: PerceptualRoughness,
    /// Range [0,1] from dielectric to pure metallic
    pub metallic_factor: f32,
    /// Specular intensity for non-metals, defaults to 0.5 on a linear scale
    /// which is mapped to 4% reflectance in the shader
    pub reflectance: f32,
    #[render_resources(ignore)]
    #[shader_def]
    pub unlit: bool,
}

impl Default for StandardMaterial {
    fn default() -> Self {
        StandardMaterial {
            base_color_factor: Color::rgb(1.0, 1.0, 1.0),
            base_color_texture: None,
            // This is the minimum the roughness is clamped to in shader code
            roughness_factor: PerceptualRoughness(0.089),
            // Few materials are purely dielectric or metallic
            // This is just a default for mostly-dielectric
            metallic_factor: 0.01,
            // Minimum real-world reflectance is 2%, most materials between 2-5%
            // Expressed in a linear scale and equivalent to 4% reflectance see https://google.github.io/filament/Material%20Properties.pdf
            reflectance: 0.5,
            unlit: false,
        }
    }
}
```

Reply:
I got some positive feedback on this on Discord, so I'm going to move forward with newtypes/enums for PBR parameters where useful.